File: apps/python-server/routers/batch.py
from datetime import datetime, timezone
from fastapi import APIRouter, Depends
from beanie import PydanticObjectId
from schemas.batch import CreateBatchRequest, UpdateBatchRequest
from utils.api_response import api_response
from middleware.auth import get_current_user, require_roles, not_student
from models.batch import Batch
from models.institution import Institution
from models.student import Student
from models.teacher import Teacher
from models.course_enrollment import CourseEnrollment
from enums import UserType

router = APIRouter(prefix="/api/v1/batches", tags=["Batches"])

_delete_roles = require_roles(UserType.SUPERADMIN, UserType.ADMIN, UserType.INSTITUTION, UserType.VENDOR)


@router.post("/create-batch")
async def create_batch(body: CreateBatchRequest, current_user: dict = Depends(not_student)):
    inst = await Institution.get(PydanticObjectId(body.institution_id))
    if not inst:
        return api_response(404, "Institution not found", error="Not found")

    batch = Batch(
        batchname=body.batchname,
        branch=body.branch,
        batch_end_year=body.batch_end_year,
        institution_id=PydanticObjectId(body.institution_id),
    )
    await batch.insert()
    return api_response(201, "Batch created", data={
        "id": str(batch.id), "batchname": batch.batchname,
        "branch": batch.branch, "batch_end_year": batch.batch_end_year,
        "institution_id": str(batch.institution_id)
    })


@router.get("/get-all-batch")
async def get_all_batches_global(current_user: dict = Depends(not_student)):
    batches = await Batch.find_all().to_list()
    data = []
    for b in batches:
        student_count = await Student.find(Student.batch_id == b.id).count()
        data.append({
            "id": str(b.id), "batchname": b.batchname, "branch": b.branch,
            "batch_end_year": b.batch_end_year, "institution_id": str(b.institution_id),
            "student_count": student_count, "created_at": b.created_at.isoformat()
        })
    return api_response(200, "Batches fetched", data=data)


@router.get("/get-all-batch/{id}")
async def get_all_batches(id: str, current_user: dict = Depends(not_student)):
    batches = await Batch.find(Batch.institution_id == PydanticObjectId(id)).to_list()
    data = []
    for b in batches:
        student_count = await Student.find(Student.batch_id == b.id).count()
        data.append({
            "id": str(b.id), "batchname": b.batchname, "branch": b.branch,
            "batch_end_year": b.batch_end_year, "institution_id": str(b.institution_id),
            "student_count": student_count, "created_at": b.created_at.isoformat()
        })
    return api_response(200, "Batches fetched", data=data)


@router.get("/get-batch-by-id/{id}")
async def get_batch_by_id(id: str, current_user: dict = Depends(get_current_user)):
    batch = await Batch.get(PydanticObjectId(id))
    if not batch:
        return api_response(404, "Batch not found", error="Not found")

    students = await Student.find(Student.batch_id == batch.id).to_list()
    student_data = [{
        "id": str(s.id), "name": s.name, "email": s.email,
        "roll_number": s.roll_number
    } for s in students]

    return api_response(200, "Batch fetched", data={
        "id": str(batch.id), "batchname": batch.batchname, "branch": batch.branch,
        "batch_end_year": batch.batch_end_year, "institution_id": str(batch.institution_id),
        "students": student_data, "created_at": batch.created_at.isoformat()
    })


@router.put("/update-batch-by-id/{id}")
async def update_batch(id: str, body: UpdateBatchRequest, current_user: dict = Depends(not_student)):
    batch = await Batch.get(PydanticObjectId(id))
    if not batch:
        return api_response(404, "Batch not found", error="Not found")

    update_data = body.model_dump(exclude_none=True)
    for key, val in update_data.items():
        setattr(batch, key, val)
    batch.updated_at = datetime.now(timezone.utc)
    await batch.save()
    return api_response(200, "Batch updated", data={"id": str(batch.id), "batchname": batch.batchname})


@router.delete("/delete-batch-by-id/{id}")
async def delete_batch(id: str, current_user: dict = Depends(_delete_roles)):
    batch = await Batch.get(PydanticObjectId(id))
    if not batch:
        return api_response(404, "Batch not found", error="Not found")

    bid = batch.id
    await Student.find(Student.batch_id == bid).delete()
    await Teacher.find(Teacher.batch_id == bid).delete()
    await CourseEnrollment.find(CourseEnrollment.batch_id == bid).delete()
    await batch.delete()
    return api_response(200, "Batch deleted")

File: apps/go-server/internal/handlers/identity/batch.go
package identity

import (
	"context"
	"net/http"
	"time"

	"github.com/labstack/echo/v4"
	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/bson/primitive"

	"github.com/bitwiselearn/go-server/internal/dbutil"
	"github.com/bitwiselearn/go-server/internal/middleware"
	"github.com/bitwiselearn/go-server/internal/models"
	"github.com/bitwiselearn/go-server/internal/response"
)

func registerBatchRoutes(g *echo.Group, am *middleware.Auth, h *handler) {
	authed := am.Required
	notStudent := am.NotStudent()
	deleteRoles := am.RequireRoles(models.RoleSuperadmin, models.RoleAdmin, models.RoleInstitution, models.RoleVendor)

	g.POST("/create-batch", h.createBatch, authed, notStudent)
	g.GET("/get-all-batch", h.getAllBatchesGlobal, authed, notStudent)
	g.GET("/get-all-batch/:id", h.getAllBatchesByInstitution, authed, notStudent)
	g.GET("/get-batch-by-id/:id", h.getBatchByID, authed)
	g.PUT("/update-batch-by-id/:id", h.updateBatch, authed, notStudent)
	g.DELETE("/delete-batch-by-id/:id", h.deleteBatch, authed, deleteRoles)
}

func (h *handler) createBatch(c echo.Context) error {
	var body createBatchRequest
	if err := c.Bind(&body); err != nil {
		return response.Err(c, http.StatusBadRequest, "Invalid request body", err.Error())
	}
	instID, ok := dbutil.ParseID(body.InstitutionID)
	if !ok {
		return response.Err(c, http.StatusNotFound, "Institution not found", "Not found")
	}
	ctx, cancel := context.WithTimeout(c.Request().Context(), reqTimeout)
	defer cancel()

	if inst, err := h.getInstitution(ctx, instID); err != nil || inst == nil {
		return response.Err(c, http.StatusNotFound, "Institution not found", "Not found")
	}

	now := time.Now().UTC()
	batch := models.Batch{
		ID: primitive.NewObjectID(), BatchName: body.BatchName, Branch: body.Branch,
		BatchEndYear: body.BatchEndYear, InstitutionID: instID, CreatedAt: now, UpdatedAt: now,
	}
	if _, err := h.coll(models.CollBatches).InsertOne(ctx, batch); err != nil {
		return response.Err(c, http.StatusInternalServerError, "Failed to create batch", err.Error())
	}
	return response.JSON(c, http.StatusCreated, "Batch created", map[string]any{
		"id": batch.ID.Hex(), "batchname": batch.BatchName, "branch": batch.Branch,
		"batch_end_year": batch.BatchEndYear, "institution_id": batch.InstitutionID.Hex(),
	}, nil)
}

func (h *handler) batchToMap(ctx context.Context, b models.Batch) map[string]any {
	studentCount := h.countColl(ctx, models.CollStudents, bson.M{"batch_id": b.ID})
	return map[string]any{
		"id": b.ID.Hex(), "batchname": b.BatchName, "branch": b.Branch,
		"batch_end_year": b.BatchEndYear, "institution_id": b.InstitutionID.Hex(),
		"student_count": studentCount, "created_at": iso(b.CreatedAt),
	}
}

func (h *handler) getAllBatchesGlobal(c echo.Context) error {
	ctx, cancel := context.WithTimeout(c.Request().Context(), reqTimeout)
	defer cancel()
	cur, err := h.coll(models.CollBatches).Find(ctx, bson.M{})
	if err != nil {
		return response.Err(c, http.StatusInternalServerError, "Failed to fetch batches", err.Error())
	}
	var batches []models.Batch
	if err := cur.All(ctx, &batches); err != nil {
		return response.Err(c, http.StatusInternalServerError, "Failed to fetch batches", err.Error())
	}
	data := make([]map[string]any, 0, len(batches))
	for _, b := range batches {
		data = append(data, h.batchToMap(ctx, b))
	}
	return response.OK(c, "Batches fetched", data)
}

func (h *handler) getAllBatchesByInstitution(c echo.Context) error {
	oid, ok := dbutil.ParseID(c.Param("id"))
	if !ok {
		return response.OK(c, "Batches fetched", []map[string]any{})
	}
	ctx, cancel := context.WithTimeout(c.Request().Context(), reqTimeout)
	defer cancel()
	cur, err := h.coll(models.CollBatches).Find(ctx, bson.M{"institution_id": oid})
	if err != nil {
		return response.Err(c, http.StatusInternalServerError, "Failed to fetch batches", err.Error())
	}
	var batches []models.Batch
	if err := cur.All(ctx, &batches); err != nil {
		return response.Err(c, http.StatusInternalServerError, "Failed to fetch batches", err.Error())
	}
	data := make([]map[string]any, 0, len(batches))
	for _, b := range batches {
		data = append(data, h.batchToMap(ctx, b))
	}
	return response.OK(c, "Batches fetched", data)
}

func (h *handler) getBatchByID(c echo.Context) error {
	oid, ok := dbutil.ParseID(c.Param("id"))
	if !ok {
		return response.Err(c, http.StatusNotFound, "Batch not found", "Not found")
	}
	ctx, cancel := context.WithTimeout(c.Request().Context(), reqTimeout)
	defer cancel()
	batch, err := h.getBatch(ctx, oid)
	if err != nil {
		return response.Err(c, http.StatusInternalServerError, "Failed to fetch batch", err.Error())
	}
	if batch == nil {
		return response.Err(c, http.StatusNotFound, "Batch not found", "Not found")
	}

	cur, err := h.coll(models.CollStudents).Find(ctx, bson.M{"batch_id": batch.ID})
	if err != nil {
		return response.Err(c, http.StatusInternalServerError, "Failed to fetch batch", err.Error())
	}
	var students []models.Student
	_ = cur.All(ctx, &students)
	studentData := make([]map[string]any, 0, len(students))
	for _, s := range students {
		studentData = append(studentData, map[string]any{
			"id": s.ID.Hex(), "name": s.Name, "email": s.Email, "roll_number": s.RollNumber,
		})
	}

	return response.OK(c, "Batch fetched", map[string]any{
		"id": batch.ID.Hex(), "batchname": batch.BatchName, "branch": batch.Branch,
		"batch_end_year": batch.BatchEndYear, "institution_id": batch.InstitutionID.Hex(),
		"students": studentData, "created_at": iso(batch.CreatedAt),
	})
}

func (h *handler) updateBatch(c echo.Context) error {
	oid, ok := dbutil.ParseID(c.Param("id"))
	if !ok {
		return response.Err(c, http.StatusNotFound, "Batch not found", "Not found")
	}
	var body updateBatchRequest
	if err := c.Bind(&body); err != nil {
		return response.Err(c, http.StatusBadRequest, "Invalid request body", err.Error())
	}
	ctx, cancel := context.WithTimeout(c.Request().Context(), reqTimeout)
	defer cancel()

	set := bson.M{"updated_at": time.Now().UTC()}
	if body.BatchName != nil {
		set["batchname"] = *body.BatchName
	}
	if body.Branch != nil {
		set["branch"] = *body.Branch
	}
	if body.BatchEndYear != nil {
		set["batch_end_year"] = *body.BatchEndYear
	}
	res, err := h.coll(models.CollBatches).UpdateOne(ctx, bson.M{"_id": oid}, bson.M{"$set": set})
	if err != nil {
		return response.Err(c, http.StatusInternalServerError, "Failed to update batch", err.Error())
	}
	if res.MatchedCount == 0 {
		return response.Err(c, http.StatusNotFound, "Batch not found", "Not found")
	}
	batch, _ := h.getBatch(ctx, oid)
	name := ""
	if batch != nil {
		name = batch.BatchName
	}
	return response.OK(c, "Batch updated", map[string]any{"id": oid.Hex(), "batchname": name})
}

func (h *handler) deleteBatch(c echo.Context) error {
	oid, ok := dbutil.ParseID(c.Param("id"))
	if !ok {
		return response.Err(c, http.StatusNotFound, "Batch not found", "Not found")
	}
	ctx, cancel := context.WithTimeout(c.Request().Context(), reqTimeout)
	defer cancel()

	batch, err := h.getBatch(ctx, oid)
	if err != nil {
		return response.Err(c, http.StatusInternalServerError, "Failed to fetch batch", err.Error())
	}
	if batch == nil {
		return response.Err(c, http.StatusNotFound, "Batch not found", "Not found")
	}
	_, _ = h.coll(models.CollStudents).DeleteMany(ctx, bson.M{"batch_id": oid})
	_, _ = h.coll(models.CollTeachers).DeleteMany(ctx, bson.M{"batch_id": oid})
	_, _ = h.coll(models.CollCourseEnrollments).DeleteMany(ctx, bson.M{"batch_id": oid})
	if _, err := h.coll(models.CollBatches).DeleteOne(ctx, bson.M{"_id": oid}); err != nil {
		return response.Err(c, http.StatusInternalServerError, "Failed to delete batch", err.Error())
	}
	return response.OK(c, "Batch deleted", nil)
}
